sdk/dotnet/Pulumi/Stack.cs

```diff
@@ -62,12 +73,20 @@ internal Stack(Func<Task<IDictionary<string, object?>>> init)
             }
         }
 
-        private async Task<IDictionary<string, object?>> RunInitAsync(Func<Task<IDictionary<string, object?>>> init)
+        internal void RegisterPropertyOutputs()
         {
-            // Ensure we are known as the root resource.  This is needed before we execute any user
-            // code as many codepaths will request the root resource.
-            await Deployment.InternalInstance.SetRootResourceAsync(this).ConfigureAwait(false);
+            var query = from property in this.GetType().GetProperties(BindingFlags.Public | BindingFlags.Instance | BindingFlags.DeclaredOnly)
+                        let attr = property.GetCustomAttribute<OutputAttribute>()
+                        where attr != null
+                        select new KeyValuePair<string, object?>(attr.Name, property.GetValue(this));
+
+            IDictionary<string, object?> outputs = new Dictionary<string, object?>(query);
```

Comment:
should we mandate that the properties are all `Output<T>`?

Reply:
I think yes, because non-Outputs cause errors down the line.

Reply:
Done